### PR TITLE
Relicensing: update copyright header to intermediate format

### PR DESCRIPTION
**Pull request by tamiko**
In reference to #16665 and #16666

### CODE REVIEW

**Comment by tamiko** [comment]
I would like to see this one merged rather sooner than later.

**Review by Rombur — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]

**Comment by tamiko** [comment]
I found yet *another* old header variant. *sigh* Cleaning up.

**Comment by masterleinad** [comment]
Since we are changing all the copyright headers anyway: Do we care about the years or should we just get rid of them as well?

**Comment by tamiko** [comment]
I like our current approach of keeping the year range as an indicator how old a file is (and when it was last modified). And we do a good job with updating copyright years as part of the release process.

But I am happy to discuss.

**Comment by masterleinad** [comment]
> I like our current approach of keeping the year range as an indicator how old a file is (and when it was last modified). And we do a good job with updating copyright years as part of the release process.

My point is that we aren't that good with setting a good initial date for files, especially in the tests where we often just copy files, and it's of limited value. We have a good approximation for the age of a file as metadata in the `git` history anyway (difficult when renaming, of course). I don't feel super strongly about it but I just wanted to bring it up also in the context of all the `Tpetra` tests that we are adding.

**Comment by tamiko** [comment]
@masterleinad Correct, the start year tends to be a bit wonky. Let me investigate whether we can improve the `update-copyright.sh` script to also grab the starting date from git history.

**Review by kronbichler — APPROVED** [review]

**Comment by tamiko** [comment]
@masterleinad I have run the updated `update-copyright.sh` script against all headers. We should now have much more precise date ranges (with some very rare exceptions when `git log --follow` wasn't able to follow a file's history completely due to less than 50% similarity between renamed file versions).

**Comment by bangerth** [comment]
From a copyright perspective, a test that is copied from another contains code that is copied and that carries the copyright years of the old test. So it isn't *always* a poor approximation of the perfect date range we should use.

**Comment by tamiko** [comment]
So... what do we want? What shall I do?

**Comment by kronbichler** [comment]
I checked the dates of some files I know the original dates (tutorials and code written initially by me or in collaboration with me), and they are pretty accurate. I like the new scheme.

**Comment by bangerth** [comment]
It seems to be a pretty universal practice to put copyright headers with years into software projects. Whether that's legally necessary is a separate question, but if it's easy enough to do, let's stick with common practice.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,17 +1,16 @@
-## ---------------------------------------------------------------------
+## ------------------------------------------------------------------------
 ##
+## SPDX-License-Identifier: LGPL-2.1-or-later
```
Comment (bangerth) [inline CMakeLists.txt:3]: I haven't seen this before, but assume it's a standardized format. Why does it only list the LGPL if we dual-license things?

Comment (bangerth) [inline CMakeLists.txt:3]: I haven't seen this before, but assume it's a standardized format. Why does it only list the LGPL if we dual-license things?

**Review by bangerth — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Comment by masterleinad** [comment]
> It seems to be a pretty universal practice to put copyright headers with years into software projects. Whether that's legally necessary is a separate question, but if it's easy enough to do, let's stick with common practice.

I just saw a couple of projects that got rid of the years in their copyrights, see, e.g., https://daniel.haxx.se/blog/2023/01/08/copyright-without-years/, and was curious what our sentiment is.

File: CMakeLists.txt (hunk 1 above)
Comment (tamiko) [reply CMakeLists.txt:3]: Yes, this identifier comes from the Linux kernel and is an attempt at unifying license header information. And it is standardized, https://spdx.org/, https://spdx.org/licenses/.

Comment (tamiko) [reply CMakeLists.txt:3]: I want to be as precise as possible: just changing the header will not magically dual license the source code. Thus, we simply state the license that effectively applies `LGPL-2.1-or-later` and point to the relicensing effort:
```
## Part of the source code is dual licensed under Apache-2.0 WITH
## LLVM-exception OR LGPL-2.1-or-later. Detailed license information
## governing the source code and code contributions can be found in
## LICENSE.md and CONTRIBUTING.md at the top level directory of deal.II.
```

**Comment by bangerth** [comment]
On 2/23/24 09:16, Matthias Maier wrote:
> +## SPDX-License-Identifier: LGPL-2.1-or-later
> 
> I want to be as precise as possible: just changing the header will not 
> magically dual license the source code. Thus, we simply state the license that 
> effectively applies |LGPL-2.1-or-later| and point to the relicensing effort:
> 
> |## Part of the source code is dual licensed under Apache-2.0 WITH ## 
> LLVM-exception OR LGPL-2.1-or-later. Detailed license information ## governing 
> the source code and code contributions can be found in ## LICENSE.md and 
> CONTRIBUTING.md at the top level directory of deal.II.|

Perhaps I misinterpret what you're trying to do. Right now, the correct 
|SPDX-License-Identifier| is indeed only |LGPL-2.1-or-later|, but that once we 
have gone through the relicensing effort, this will become |Apache-2.0 WITH ##
 > LLVM-exception OR LGPL-2.1-or-later| following the example at 
https://spdx.dev/learn/handling-license-info/ ?

I was mainly confused by the mismatch between the SPDX-License-Identifier 
listing only one license, and our end goal of the dual licensing. If there is 
an area where I would like to avoid any ambiguity, it's the licensing part :-)

**Comment by tamiko** [comment]
@bangerth Yes. The moment we have sufficient signatures that we are indeed covered by the dual-license to a sufficient degree we will have to switch the headers again.

**Comment by bangerth** [comment]
On 2/23/24 14:58, Matthias Maier wrote:
> @bangerth <https://github.com/bangerth> Yes. The moment we have sufficient 
> signatures that we are indeed covered by the dual-license to a sufficient 
> degree we will have to switch the headers again.

Alright, thanks for confirming!

**Pull request merged**